doc/defining.md

```diff
@@ -0,0 +1,272 @@
+type Mock interface {
+    // Component methods are included in Mock.
+    Component
+
+    // AddedFoos returns the foos added by AddFoo calls on the mock implementation.
+    AddedFoos() []Foo
+}
+
+// MockModule defines the fx options for the mock component.
+var MockModule = fx.Module(
+    "comp/foo",
+    fx.Provide(newMockFoo),
+)
+```
+
+The `newMockFoo` constructor should create an implementation of the Mock interface.
+
+## Defining a Bundle
+
+A bundle is defined in a dedicated package name `comp/<bundlename>`.
+The package must have the following defined in `bundle.go`:
+
+ * Extensive package-level documentation.
+   This should define:
+
+     * The purpose of the bundle
+     * What components are and are not included in the bundle.
+       Components might be omitted in the interest of binary size, as discussed in the [overview](./components.md).
+     * Which components are automatically instantiated.
+     * Which other _bundles_ this bundle depends on.
+       Bundle dependencies are always expressed at a bundle level.
+
+ * A team-name comment of the form `// team: <teamname>`.
+   This is used to generate CODEOWNERS information.
+
+ * `componentName` -- the Go path of the component, relative to the repository root, e.g., `comp/core/health`.
+
+ * `BundleParams` -- the type of the bundle's parameters (see below).
+   This item should have a formulaic doc string like `// BundleParams defines the parameters for this bundle.`
+
+ * `Bundle` -- an `fx.Option` that can be included in an `fx.App` to make this bundle's components available.
+   To assist with debugging, use `fx.Module(componentName, ..)`.
+   Use `fx.Invoke(func(componentpkg.Component) {})` to instantiate components automatically.
+   This item should have a formulaic doc string like `// Module defines the fx options for this component.`
+
+Typically, a bundle will automatically instantiate the top-level components that represent the bundle's purpose.
+For example, the trace-agent bundle `comp/trace` might automatically instantiate `comp/trace/agent`.
+
+### Bundle Parameters
+
+Apps can provide some intialization-time parameters to bundles.
+These parameters are limited to two kinds:
+
+ * Parameters specific to the app, such as whether to start a network server; and
+ * Parameters from the environment, such as command-line options.
+
+Anything else is runtime configuration and should be handled vi `comp/core/config` or another mechanism.
+
+To avoid Go package cycles, the `BundleParams` type must be defined in the bundle's internal package, and re-exported from the bundle package:
+
+```go
+// --- comp/<bundlename>/internal/params.go ---
+
+// BundleParams defines the parameters for this bundle.
+type BundleParams struct {
+    ...
+}
+
+// --- comp/<bundlename>/bundle.go ---
+import ".../comp/<bundlename>/internal"
+// ...
+
+// BundleParams defines the parameters for this bundle.
+type BundleParams = internal.BundleParams
+```
```

Comment:
Would be good to split this into two code blocks to explicitly show that we are talking about two different files